From 133a54825609acff6b8eb3e71a4d07287fdeb980 Mon Sep 17 00:00:00 2001
From: Markus Mayer <mmayer@broadcom.com>
Date: Mon, 13 May 2019 11:22:43 -0700
Subject: [PATCH] tftp: introduce CONFIG_FEATURE_TFTP_UMASK

We make the umask configurable that TFTP is using. This way it is easy
to make files transferred by TFTP executable if that is desired by the
user.

Signed-off-by: Markus Mayer <mmayer@broadcom.com>
---
 networking/tftp.c | 15 +++++++++++++--
 1 file changed, 13 insertions(+), 2 deletions(-)

diff --git a/networking/tftp.c b/networking/tftp.c
index f5b4367ca9a4..2758fefc3850 100644
--- a/networking/tftp.c
+++ b/networking/tftp.c
@@ -75,6 +75,17 @@
 //config:	Allow tftp to specify block size, and tftpd to understand
 //config:	"blksize" and "tsize" options.
 //config:
+//config:config FEATURE_TFTP_UMASK
+//config:	hex "UMASK for TFTP client and server to use"
+//config:	default 0x1b6
+//config:	depends on TFTP || TFTPD
+//config:	help
+//config:	Specify the UMASK to be used for new files being transferred
+//config:	via TFTP. Defaults to 0x1b6 (0666). Use 0x1ff (0777) if you
+//config:	want execute permissions on transferred files.
+//config:	(Must be specified in hex, since Kconfig doesn't support octal
+//config:	fields.)
+//config:
 //config:config TFTP_DEBUG
 //config:	bool "Enable debug"
 //config:	default n
@@ -394,7 +405,7 @@ static int tftp_protocol(
 				open_mode = O_WRONLY | O_TRUNC;
 			}
 		}
-		local_fd = open(local_file, open_mode, 0666);
+		local_fd = open(local_file, open_mode, CONFIG_FEATURE_TFTP_UMASK);
 		if (local_fd < 0) {
 			/* sanitize name, it came from untrusted remote side */
 			unsigned char *p = (void *) local_file;
@@ -430,7 +441,7 @@ static int tftp_protocol(
 		if (CMD_PUT(option_mask32)) {
 			local_fd = STDIN_FILENO;
 			if (local_file)
-				local_fd = xopen(local_file, O_RDONLY);
+				local_fd = xopen3(local_file, O_RDONLY, CONFIG_FEATURE_TFTP_UMASK);
 		}
 /* Removing #if, or using if() statement instead of #if may lead to
  * "warning: null argument where non-null required": */
-- 
2.25.1

